src/librustc_typeck/check/mod.rs

```diff
@@ -450,13 +450,36 @@ pub enum Diverges {
 
     /// Definitely known to diverge and therefore
     /// not reach the next sibling or its parent.
-    Always,
+    Always {
+        /// The `Span` points to the expression
+        /// that caused us to diverge
+        /// (e.g. `return`, `break`, etc).
+        span: Span,
+        /// In some cases (e.g. a `match` expression
+        /// where all arms diverge), we may be
+        /// able to provide a more informative
+        /// message to the user.
+        /// If this is `None`, a default messsage
+        /// will be generated, which is suitable
+        /// for most cases.
+        custom_note: Option<&'static str>
+    },
 
     /// Same as `Always` but with a reachability
     /// warning already emitted.
     WarnedAlways
 }
 
+impl Diverges {
+    /// Creates a `Diverges::Always` with the provided span and the default note message
```

Comment:
```suggestion
    /// Creates a `Diverges::Always` with the provided `span` and the default note message.
```

Also, can you merge with the impl of `is_always`?

Reply:
I deliberately specified both fields in `is_always`, to make it clear that they have dummy values.

Reply:
I don't follow; my comment was regarding having two `impl` blocks instead of co-locating the functions. :)

Reply:
Oh - I assumed that you meant using the `always` constructor in `is_always`.